torch/_inductor/runtime/triton_helpers.py

```diff
@@ -614,3 +614,52 @@ def select_one(x, mask, dim, keep_dims=False):
     ix = x.to(idtype, bitcast=True)
     iy = tl.sum(ix * mask, dim, keep_dims=keep_dims)
     return iy.to(x.dtype, bitcast=True)
+
+
+@triton.jit
+def gpu_barrier(
+    sem, total: tl.constexpr, first_barrier: tl.constexpr = tl.constexpr(False)
+):
+    """
+    Wait for all other thread blocks to reach this barrier before returning.
+
+    if *sem <= total:   we are in count_up mode waiting for all threads to enter
+    if *sem >= total*2: we are in count_down mode waiting for all threads to leave
+
+    Args:
+        sem: an int32 semaphores, zero initialized
+        total: how many other blocks are running
+        first_barrier: True if we can skip waiting for exit from last barrier.  Can set either on the first call,
+                       or if one ping-pongs between two semaphores.
+    """
+    # ensure stores before this are visible
+    tl.debug_barrier()
+
+    # scaling factor so we know which mode we are in
+    count_down = 2 * total
+
+    if not first_barrier:
+        # all threads must exit prior barrier before we start
```

Comment:
with the approach above you don't need to wait for threads to exit prior barrier - even if some threads are still lingering, the new threads entering second barrier won't be able to flip the signal